pgxn/neon/neon.c

```diff
@@ -151,13 +190,87 @@ LogicalSlotsMonitorMain(Datum main_arg)
 					if (slots[i].dropped)
 					{
 						elog(LOG, "Drop logical replication slot because it was not update more than %ld seconds",
-							 (now - slots[i].last_updated)/USECS_PER_SEC);
+							 (now - slots[i].last_updated) / USECS_PER_SEC);
 						ReplicationSlotDrop(slots[i].name.data, true);
 						slots[i].dropped = false;
 					}
 				}
 			}
 		}
+
+		/*
+		 * If there are too many .snap files, just drop all logical slots to
+		 * prevent aux files bloat.
+		 */
+		num_snap_files = get_num_snap_files();
+		if (num_snap_files > logical_replication_max_snap_files)
+		{
+			elog(LOG, "ls_monitor: dropping all logical slots: found %d .snap files, limit is %d",
+				 num_snap_files, logical_replication_max_snap_files);
+			for (int i = 0; i < max_replication_slots; i++)
+			{
+				char		slot_name[NAMEDATALEN];
+				ReplicationSlot *s = &ReplicationSlotCtl->replication_slots[i];
+
+				/* find the name */
+				LWLockAcquire(ReplicationSlotControlLock, LW_SHARED);
```

Comment:
it is better to move this lock outside the loop

Reply:
no, because drop takes this lock in excl mode

Reply:
What I am  suggesting:
```
			LWLockAcquire(ReplicationSlotControlLock, LW_SHARED);
			for (int i = 0; i < max_replication_slots; i++)
			{
				char		slot_name[NAMEDATALEN];
				ReplicationSlot *s = &ReplicationSlotCtl->replication_slots[i];

				/* find the name */
				/* Consider only logical repliction slots */
				if (!s->in_use || !SlotIsLogical(s))
				{
					continue;
				}
				strlcpy(slot_name, s->data.name.data, NAMEDATALEN);
				elog(LOG, "ls_monitor: dropping slot %s", slot_name);
				LWLockRelease(ReplicationSlotControlLock);
				...
				LWLockAcquire(ReplicationSlotControlLock, LW_SHARED);
			}
			LWLockRelease(ReplicationSlotControlLock);
```
because most slots are not used. 
But it is not principle as far as `max_replication_slots` is small (default valuer is 10).